server/controller/src/main/java/ai/starwhale/mlops/schedule/impl/docker/RunExecutorDockerImpl.java

```diff
@@ -55,28 +57,44 @@ public class RunExecutorDockerImpl implements RunExecutor {
     final ExecutorService cmdExecThreadPool;
     final ContainerRunMapper containerRunMapper;
     final String nodeIp;
-    final String network;
+    final String networkName;
+
+    Network network;
 
     public RunExecutorDockerImpl(
             DockerClientFinder dockerClientFinder,
             HostResourceConfigBuilder hostResourceConfigBuilder,
             ExecutorService cmdExecThreadPool,
             ContainerRunMapper containerRunMapper,
             String nodeIp,
-            String network
+            String networkName
     ) {
         this.dockerClientFinder = dockerClientFinder;
         this.hostResourceConfigBuilder = hostResourceConfigBuilder;
         this.cmdExecThreadPool = cmdExecThreadPool;
         this.containerRunMapper = containerRunMapper;
         this.nodeIp = nodeIp;
-        this.network = network;
+        this.networkName = networkName;
     }
 
     @Override
     public void run(Run run, RunReportReceiver runReportReceiver) {
 
         DockerClient dockerClient = dockerClientFinder.findProperDockerClient(run.getRunSpec().getResourcePool());
+        if (null == network) {
+            synchronized (this) {
+                if (null == network) {
+                    try {
+                        network = dockerClient.inspectNetworkCmd().withNetworkId(this.networkName).exec();
+                    } catch (NotFoundException e) {
+                        dockerClient.createNetworkCmd().withName(this.networkName).exec();
+                        network = dockerClient.inspectNetworkCmd().withNetworkId(this.networkName).exec();
+                    }
+                }
+            }
+        }
```

Comment:
> https://codeql.github.com/codeql-query-help/java/java-unsafe-double-checked-locking/
> 
> Just ensure the network in the constructor?

- First let's talk about the article you quote. This article argues that double checked locking is not thread safe because of returning partially initiallized object or memory inconsistent object. It gives three examples, two of which returns  partially initiallized object . The last example returns inconsitent object. But this won't happen in our case. Quote from [oracle](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility):

> Each action in a thread happens-before every action in that thread that comes later in the program's order.

This rule guarentees that Line94 happens before Line95, which means `network` is fetched after the network creating action successfully returned.
https://github.com/star-whale/starwhale/blob/525e311d953b318dc0ffba62916846d4f465c9f1/server/controller/src/main/java/ai/starwhale/mlops/schedule/impl/docker/RunExecutorDockerImpl.java#L94-L95

> An unlock (synchronized block or method exit) of a monitor happens-before every subsequent lock (synchronized block or method entry) of that same monitor. And because the happens-before relation is transitive, all actions of a thread prior to unlocking happen-before all actions subsequent to any thread locking that monitor.

This rule guarentees that the initialization of `network` happens before all subsequent threads that reads `network`
https://github.com/star-whale/starwhale/blob/525e311d953b318dc0ffba62916846d4f465c9f1/server/controller/src/main/java/ai/starwhale/mlops/schedule/impl/docker/RunExecutorDockerImpl.java#L88-L99

- Second , why not put it in the constructor. Because the 'dockerClient' can only be got in `run` method within current framework. Change the framework is another issue :) 
https://github.com/star-whale/starwhale/blob/525e311d953b318dc0ffba62916846d4f465c9f1/server/controller/src/main/java/ai/starwhale/mlops/schedule/impl/docker/RunExecutorDockerImpl.java#L84-L86